File: 1510-stone-game-iv/1510-stone-game-iv.cpp
class Solution {
public:
    bool winnerSquareGame(int n) {
        vector<bool>dp(n+1);
        // dp[0]=0;
        // dp[1]=1;
        for(int i=0;i<=n;i++){
            if(dp[i]) continue;
            for(int j=1;i+j*j<=n;j++){
                // cout<<dp[i-j*j]<<" "<<i<<" "<<j<<endl;
                dp[i+j*j]=true;
            }
        }
        return dp[n];
    }
};

File: 1510-stone-game-iv/1510-stone-game-iv.java
class Solution {
    public boolean winnerSquareGame(int n) {
        HashMap<Integer,Boolean>cache=new HashMap<>();
        cache.put(0,false);
        return dfs(cache,n);
    }
    public static boolean dfs(HashMap<Integer,Boolean>cache,int remain){
        if(cache.containsKey(remain))
            return cache.get(remain);
        int sqrt=(int)Math.sqrt(remain);
        for(int i=1;i<=sqrt;i++){
            if(!dfs(cache,remain-i*i)){
                cache.put(remain,true);
                return true;
            }
        }
        cache.put(remain,false);
        return false;
    }
}